### PR TITLE
Fix selective masking composition stage 

### LINKED ISSUE #34650: Selective masking: phantom lines on dark background
**Describe the bug**

Selective masking generates lines or "halos" around the masked area. They are visible on dark backgrounds.

Some examples:

image

image

image

When turning off anti-aliasing, the problem disappears, but this is not a solution.
Something is wrong in the way images are composed and leads to this rendering artifacts

### LINKED ISSUE #34947: Selective masking: issue with mask opacity
**Describe the bug**

Playing with the opacity of selective masking sometimes give strange results.

Initially reported by @nirvn https://github.com/qgis/QGIS/pull/33432#issuecomment-566422039 and on Twitter : https://twitter.com/klaskarlsson/status/1231518713804992512

image

image

**How to Reproduce**

I fail to reproduce the error on my system. But since two people reported it, there might be something ...

**QGIS and OS versions**

3.12

### PR DESCRIPTION
**Pull request by obrix**
At the composition stage of the second pass introduced with selective masking, the unmasked elements located under the mask are copied into a temporary image buffer and then re added later to the final image.

This copy was done using the masks from selective masking with their full opacity informations. Unmasked informations were thus combined with opacity from the mask leading to the introduction of several artefacts : 
* The "phantom lines" #34650 (opacity information introduced by the antialising on mask drawing)
* Opacity applied on non masked surfaces in #34947

To fix this issue a temporary mask is created for the copy of unmasked pixels, in this temporary mask the alpha channel is "binarized" (fully opaque on masking pixel, fully transparent on non masking pixel).

Should fix #34650 and fix #34947

Here are two examples of two issues fixed by this PR

Phantom lines : 
phantom_lines

Phantom lines fixed : 
phantom_lines_fix

Opacity applied on yellow line that is not masked : 
opacity_issue

Opacity fixed (applied only on masked black line) : 
opacity_issue_fix

### CODE REVIEW

**Review by nyalldawson — COMMENTED** [review]
This is going to be sloooow. Could the original image be constructed using the QImage::Format_Mono to avoid this requirement?

**Comment by nirvn** [comment]
@obrix , great to see somehow attempting to fix this issue :+1:

**Comment by obrix** [comment]
@nyalldawson Not sure about Format_Mono, I have to look how this would interact with QPainter CompositionMode. I made sure to avoid using setPixel to build my image but there is probably a faster way to do this, I'm not an expert on QImage. @mhugo Do you have any idea ?

@nirvn Thanks : )

**Comment by mhugo** [comment]
@nyalldawson 's suggestion seems good: it would be better to already construct a 1-bit mask rather than trying to convert it afterwards.

**Comment by obrix** [comment]
@nyalldawson @mhugo Masks with an alpha channel must be kept to handle the transparency features of selective masking correctly. So this means duplicating all masks (LayerRenderJob::maskImage and LabelRenderJob::maskImages) with 1-bit versions of the same image. Duplicating painter draw calls too. It introduces some more complexity to the code and I am not sure the performances will be better this way (but maybe I am wrong) ?

After some tests it seems converting the final mask to 1-bpp mask using QImage::createAlphaMask instead of iterating on the pixel is way better performance wise. Just need to find a way to compose this 1-bpp mask in the qpainter to get the same result as the current "CompositionMode_DestinationIn" composition.

**Comment by nyalldawson** [comment]
@obrix another option could be calling QImage::convertToFormat on the mask image and convert it to mono using the optimised qt format conversion method instead of the handcoded version.

On another note -- please tell me your TODO list includes refactoring masking to make it vector - export friendly! The forced rasterisation is such a huge limitation which prevents this feature being used for any serious work.

**Comment by obrix** [comment]
@nyalldawson I implemented a faster method to compute my mask, this should be better performance wise. I also slightly modified the composition again to fix some issues I had with the composition of effects. The tests are now good.
Concerning the refactoring to make the feature vector friendly, as you saw this is the subject of a recent QEP by @mhugo  that I may realize !

**Comment by nyalldawson** [comment]
> Concerning the refactoring to make the feature vector friendly, as you saw this is the subject of a recent QEP by @mhugo that I may realize !

Yep, saw that :+1: 

If you're redesigning how this feature works, I think you should also consider https://github.com/qgis/QGIS/issues/34352 -- that one also points to an architectural limitation in the current approach...

**Comment by andreasneumann** [comment]
> If you're redesigning how this feature works, I think you should also consider #34352 -- that one also points to an architectural limitation in the current approach...

@nyalldawson - yes this is also on the radar and I asked the original funders to finance solving this issue. Slowly we should be getting there ;-) selective masking is a complex new feature and it turned out that it needs additional work. Thank you also for helping us getting there!

File: src/core/qgsmaprendererjob.cpp [hunk 1]
```diff
@@ -817,7 +817,17 @@ void QgsMapRendererJob::composeSecondPass( LayerRenderJobs &secondPassJobs, Labe
       // Only retain parts of the second rendering that are "inside" the mask image
       QPainter *painter = job.context.painter();
       painter->setCompositionMode( QPainter::CompositionMode_DestinationIn );
-      painter->drawImage( 0, 0, *maskImage );
+
+      //Create an "alpha binarized" image of the maskImage to :
+      //* Eliminate antialiasing artefact
+      //* Avoid applying mask opacity to elements under the mask but not masked
+      QImage maskBinAlpha = maskImage->createMaskFromColor( 0 );
+      maskBinAlpha.setColorCount( 0 );
```
Comment (nyalldawson) [inline src/core/qgsmaprendererjob.cpp:825]: Is this necessary? Seems like it would be redundant given the call to setColorTable below...

**Comment by nyalldawson** [comment]
Looking good!

File: src/core/qgsmaprendererjob.cpp (hunk 1 above)
Comment (obrix) [reply src/core/qgsmaprendererjob.cpp]: I had some issues at one point with color count being incremented but you are right it seems it is now useless. I removed it.

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**